File: leetcode/cpp/954. 二倍数对数组.cpp
#include <bits/stdc++.h>

using namespace std;

bool canReorderDoubled(vector<int>& arr) {
    unordered_map<int, int> cnt;
    for (auto i : arr) {
        cnt[i]++;
    }
    sort(arr.begin(), arr.end(), [](int i, int j) {
        return abs(i) < abs(j);
    });
    for (auto &i : arr) {
        if (cnt[i] != 0 && cnt.find(2 * i) != cnt.end()) {
            cnt[i]--;
            cnt[2 * i]--;
        }
    }
    for (auto &kv : cnt) {
        if (kv.second != 0) {
            return false;
        }
    }
    return true;
}

int main() {
    vector<int> arr({2,1,2,1,1,1,2,2});
    cout << canReorderDoubled(arr) << endl;

    return 0;
}

File: leetcode/golang/problems/954. 二倍数对数组.go
package problems

import (
	"sort"
)

func CanReorderDoubled(arr []int) bool {
	sort.Slice(arr, func(i, j int) bool {
		return abs(arr[i]) < abs(arr[j])
	})
	cnt := map[int]int{}
	for _, v := range arr {
		cnt[v]++
	}
	if cnt[0] % 2 != 0 {
		return false
	}
	for _, v := range arr {
		if _, ok := cnt[2*v]; ok && cnt[v] != 0 {
			cnt[v]--; cnt[2*v]--
		}
	}
	for _, v :=range cnt {
		if v != 0 {
			return false
		}
	}

	return true
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}